### ISSUE TITLE
Exception calling "FindAll" with "0" argument<s>

### ISSUE THREAD

**Issue by EHermanns** [BloodHoundAD/BloodHound#39]
Hello,

I'm attempting to run Bloodhound on an internal test domain separate from my own, however when I run:
`Get-BloodHoundData -Domain "FakeDomainName.com" | Export-BloodHoundCSV`

I get an error stating that 
`Exception calling "FindAll" with "0" argument(s): "A referral was returned from the server.
"
At C:\users\fakeUser\Git\Bloodhound\PowerShell\BloodHound.ps1:5232 char:50
                 $Results = $GroupSearcher.FindAll <<<< ()
     CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
     FullyQualifiedErrorId : DotNetMethodException`

`You cannot call a method on a null-valued expression.
At C:\users\fakeUser\Git\Bloodhound\PowerShell\BloodHound.ps1:5246 char:33
                 $Results.dispose <<<< ()
     CategoryInfo          : InvalidOperation: (dispose:String) [], RuntimeException
     FullyQualifiedErrorId : InvokeMethodOnNull`

`WARNING: Error: Exception calling "FindAll" with "0" argument(s): "A referral was returned from the server."`

The command stops here for stops here for at least a few minutes, no more than 10 before continuing
`Exception calling "FindAll" with "0" argument(s): "A referral was returned from the server.
"
At C:\users\fakeUser\Git\Bloodhound\PowerShell\BloodHound.ps1:13415 char:61
                 ForEach($UserResult in $UserSearcher.FindAll <<<< ()) {
     CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
     FullyQualifiedErrorId : DotNetMethodException
WARNING: Error: Exception calling "FindAll" with "0" argument(s): "A referral was returned from the server."`

`Invoke-ThreadedFunction : Cannot bind argument to parameter 'ComputerName' because it is null.
At C:\users\fakeUser\Git\Bloodhound\PowerShell\BloodHound.ps1:13624 char:89
                     $TargetComputers2 = Invoke-ThreadedFunction -NoImports -ComputerName <<<<  $TargetComputers -Scri
ptBlock $Ping -Threads 100
     CategoryInfo          : InvalidData: (:) [Invoke-ThreadedFunction], ParameterBindingValidationException
     FullyQualifiedErrorId : ParameterArgumentValidationErrorNullNotAllowed,Invoke-ThreadedFunction`

`Get-NetSession : Cannot validate argument on parameter 'ComputerName'. The argument is null or empty. Supply an argumen
t that is not null or empty and then try the command again.
At C:\users\fakeUser\Git\Bloodhound\PowerShell\BloodHound.ps1:13521 char:61
                     $Sessions = Get-NetSession -ComputerName <<<<  $ComputerName
     CategoryInfo          : InvalidData: (:) [Get-NetSession], ParameterBindingValidationException
     FullyQualifiedErrorId : ParameterArgumentValidationError,Get-NetSession`

`Get-NetLoggedon : Cannot validate argument on parameter 'ComputerName'. The argument is null or empty. Supply an argume
nt that is not null or empty and then try the command again.
At C:\users\fakeUser\Git\Bloodhound\PowerShell\BloodHound.ps1:13556 char:62
                     $LoggedOn = Get-NetLoggedon -ComputerName <<<<  $ComputerName
     CategoryInfo          : InvalidData: (:) [Get-NetLoggedon], ParameterBindingValidationException
     FullyQualifiedErrorId : ParameterArgumentValidationError,Get-NetLoggedon`

`Get-LoggedOnLocal : Cannot validate argument on parameter 'ComputerName'. The argument is null or empty. Supply an argu
ment that is not null or empty and then try the command again.
At C:\users\fakeUser\Git\Bloodhound\PowerShell\BloodHound.ps1:13579 char:69
                     $LocalLoggedOn = Get-LoggedOnLocal -ComputerName <<<<  $ComputerName
     CategoryInfo          : InvalidData: (:) [Get-LoggedOnLocal], ParameterBindingValidationException
     FullyQualifiedErrorId : ParameterArgumentValidationError,Get-LoggedOnLocal`

The end result is the creation of two CSV's (local_admins and trusts) that do have some data on them, but I have no idea if those are complete or not, I haven't worked with this domain and have no idea what's there

I have practically no experience with AD and PowerShell, so if it's a rather simple fix that is extremely well known, I apologize in advance for not knowing it and taking up your time. Also, apologies for the terrible code segments. The preview was giving me issues with some of the characters that were in the error, and formatting has never been a forte of mine.

Thanks for any and all help!

**Comment by HarmJ0y** [comment]
Thanks, we'll try to check into this soon. Some of the errors describe "A referral was returned from the server", which often means the target DC returned a referral for another domain's DC (http://stackoverflow.com/questions/6954170/a-referral-was-returned-from-the-server-exception-when-accessing-ad-from-c-sha). Does your test domain have any trusts present? (this also could be/is likely a bug on our side)

**Comment by EHermanns** [comment]
Sorry for the delay in replying.
From what I've been told, there is a one way trust, though I'm not sure if that's my domain to target domain, target domain to my domain, etc.

**Comment by HarmJ0y** [comment]
FYI we just tuned up the ingestor heavily on this branch - https://github.com/adaptivethreat/BloodHound/blob/ingestor_tuneup/PowerShell/BloodHound.ps1 , so might be worth another try to see if our increased error handling helps.

All the enumeration and export logic was combined into Invoke-BloodHound -> by default it will output to csvs in the current directory, using -URI/-UserPass will switch to the neo4j REST ingestion instead. -CollectionMethod Stealth was heavily fixed/tuned as well.

**Comment by EHermanns** [comment]
Exception calling "FindOne" with "0" argument(s): "A referral was returned from the server.
"
At C:\users\fakeuser\git\BloodHound-ingestor_tuneup\BloodHound-ingestor_tuneup\PowerShell\BloodHound.ps1:2313 char:40
+     $Result = $ComputerSearcher.FindOne <<<< ()
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : DotNetMethodException`

Exception calling "FindAll" with "0" argument(s): "A referral was returned from the server.
"
At C:\users\fakeuser\git\BloodHound-ingestor_tuneup\BloodHound-ingestor_tuneup\PowerShell\BloodHound.ps1:4840 char:63
+                 ForEach($GroupResult in $GroupSearcher.FindAll <<<< ()) {
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : DotNetMethodException

Exception calling "FindOne" with "0" argument(s): "A referral was returned from the server.
"
At C:\users\fakeuser\git\BloodHound-ingestor_tuneup\BloodHound-ingestor_tuneup\PowerShell\BloodHound.ps1:2313 char:40
+     $Result = $ComputerSearcher.FindOne <<<< ()
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : DotNetMethodException

Exception calling "FindAll" with "0" argument(s): "A referral was returned from the server.
"
At C:\users\fakeuser\git\BloodHound-ingestor_tuneup\BloodHound-ingestor_tuneup\PowerShell\BloodHound.ps1:5286 char:73
+                     ForEach($ComputerResult in $ComputerSearcher.FindAll <<<< ()) {
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : DotNetMethodException

Unfortunately, no such luck. I'm beginning to think it has more to do with how the AD is set up, as opposed to BloodHound.

**Comment by mbrancato** [comment]
I can confirm that I experience this similar issue in a large child domain. I'm guessing there are a lot of trusts as there are numerous other child domains. I run into similar issues with the ingestor in 'master'.

`An error occurred while enumerating through a collection: An operations error o
ccurred.
.
At line:4482 char:12  
+     ForEach <<<< ($User in $UserSearcher.FindAll()) {
    + CategoryInfo          : InvalidOperation: (System.Director...sultsEnumer
   ator:ResultsEnumerator) [], RuntimeException
    + FullyQualifiedErrorId : BadEnumeration

Exception calling "FindOne" with "0" argument(s): "An operations error occurred
.
"
At line:2313 char:40  
+     $Result = $ComputerSearcher.FindOne <<<< ()
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : DotNetMethodException

Exception calling "FindAll" with "0" argument(s): "An operations error occurred
.
"
At line:4846 char:40  
+                 $ObjectSearcher.FindAll <<<< () | ForEach-Object {
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : DotNetMethodException

Exception calling "FindAll" with "0" argument(s): "An operations error occurred
.
"
At line:3082 char:47  
+     ForEach($GPOResult in $GPOSearcher.FindAll <<<< ()) {
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : DotNetMethodException

Exception calling "FindOne" with "0" argument(s): "An operations error occurred
.
"
At line:2313 char:40  
+     $Result = $ComputerSearcher.FindOne <<<< ()
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : DotNetMethodException

Exception calling "FindAll" with "0" argument(s): "An operations error occurred
.
"
At line:2400 char:84  
+     Sort-Object -Unique -InputObject $(ForEach($UserResult in $UserSearcher.F
indAll <<<< ()) {if($UserResult.Properties['homedirectory']) {Split-Path($UserR
esult.Properties['homedirectory'])}if($UserResult.Properties['scriptpath']) {Sp
lit-Path($UserResult.Properties['scriptpath'])}if($UserResult.Properties['profi
lepath']) {Split-Path($UserResult.Properties['profilepath'])}})
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : DotNetMethodException

WARNING: Get-DFSshareV1 error : Exception calling "FindAll" with "0"
argument(s): "An operations error occurred.
"
WARNING: Get-DFSshareV2 error : Exception calling "FindAll" with "0"
argument(s): "An operations error occurred.
"
WARNING: Error: Exception calling "FindAll" with "0" argument(s): "An
operations error occurred.
"
`

**Comment by mbrancato** [comment]
I've submitted pull request #56 as a possible solution. I'm not entire sure of the cause, but Get-GlobalCatalogUserMapping would fail immediately when invoked via WMI exec with the following

> An error occurred while enumerating through a collection: An operations error o
> ccurred.
> .
> At line:4482 char:12
> +     ForEach <<<< ($User in $UserSearcher.FindAll()) {
>     + CategoryInfo          : InvalidOperation: (System.Director...sultsEnumer
>    ator:ResultsEnumerator) [], RuntimeException
>     + FullyQualifiedErrorId : BadEnumeration

An interactive PowerShell session and trying to .Count all the the users would cause the system to use a bunch of memory and take forever. I decided to bypass trusts and GC deconfliction  to attempt to only look at relationships in the current domain. This appeared to work but I'm unsure of the consequences of doing this and how it relates to the other information gathered. I tried it in one domain and received good results.
